src/markov/ddp.jl

```diff
@@ -0,0 +1,469 @@
+    v::Vector{Float64}
+    Tv::Array{Float64}
+    num_iter::Int
+    sigma::Array{Int,1}
+    mc::MarkovChain
+
+    function DPSolveResult(ddp::DiscreteDP)
+        v = s_wise_max(ddp.R)                         #Initialise v with v_init
+        ddpr = new(v, similar(v), 0, similar(v, Int))
+
+        # fill in sigma with proper values
+        compute_greedy!(ddp, ddpr)
+        ddpr
+    end
+
+    # method to pass initial value function (skip the s-wise max)
+    function DPSolveResult(ddp::DiscreteDP, v::Vector)
+        ddpr = new(v, similar(v), 0, similar(v, Int))
+
+        # fill in sigma with proper values
+        compute_greedy!(ddp, ddpr)
+        ddpr
+    end
+end
+
+#-----------------#
+#-Inner Functions-#
+#-----------------#
+
+"""
+Return the vector max_a vals(s, a), where vals is represented
+by a 2-dimensional ndarray of shape (self.num_states,
+self.num_actions). Stored in out, which must be of length
+self.num_states.
+out and out_argmax must be of length self.num_states; dtype of
+out_argmax must be int.
+
+"""
+s_wise_max(vals::Matrix) = vec(maximum(vals, 2))
+s_wise_max!(vals::Matrix, out::Vector) = maximum!(out, vals)
+s_wise_max!{T}(vals::Matrix{T}, out::Vector, out_argmax) =
+    Base.findminmax!(Base.MoreFun(), fill!(out, -Inf) , out_argmax, vals)
+
+#=
+Define Matrix Multiplication between 3-dimensional matrix and a vector
+
+Matrix multiplication over the last dimension of A
+
+=#
+function *{T}(A::Array{T,3}, v::Vector)
+    size(A, 3) ==  size(v, 1) || error("wrong dimensions")
+    out = Array(T, size(A)[1:2])
+
+    # TODO: slicing A[i, j, :] is cache-unfriendly
+    for j=1:size(out, 2), i=1:size(out, 1)
+        out[i, j] = dot(A[i, j, :][:], v)
+    end
+    out
+end
+
+"""
+The Bellman operator, which computes and returns the updated value function Tv
+for a value function v.
+
+##### Parameters
+
+- `ddp::DisreteDP` : Object that contains the model parameters
+- `ddpr::DPSolveResult` : Object that contains result variables
+
+##### Returns
+
+- `ddpr.Tv::Array{Float64,1}` : Updated value function vector
+
+##### Notes
+
+Updates `ddpr.Tv` and `ddpr.sigma` inplace
+
+"""
+function bellman_operator!(ddp::DiscreteDP, ddpr::DPSolveResult)
+    vals = ddp.R + ddp.beta * ddp.Q * ddpr.v
+    s_wise_max!(vals, ddpr.Tv, ddpr.sigma)
+    ddpr.Tv
+end
+
+"""
+The Bellman operator, which computes and returns the updated value function Tv
+for a given value function v.
+
+This function will fill the input `v` with `Tv` and the input `sigma` with the
+corresponding policy rule
+
+##### Parameters
+
+- `ddp::DisreteDP`: The ddp model
+- `v::Vector{Float64}`: The current guess of the value function
+- `sigma::Vector{Int}`: A buffer array to hold the policy function. Initial
+  values not used and will be overwritten
+
+##### Returns
+
+- `Tv::Vector{Float64}`: Updated value function vector
+- `sigma::Vector{Float64}`: Policy rule
+"""
+function bellman_operator!(ddp::DiscreteDP, v::Vector{Float64}, sigma::Vector{Int})
```

Comment:
When I tried, `v = [0, 0, 0]` wasn't accepted, but had to do `v = [0., 0., 0.]`. Would it be possible to allow `v` to be any `Real`?